include/range/v3/utility/swap.hpp

```diff
@@ -42,10 +53,26 @@ namespace ranges
         /// \cond
         namespace adl_swap_detail
         {
-            using std::swap;
+
+            // The following implementation of swap is adapted from libc++:
+            template <class T>
+            RANGES_CXX14_CONSTEXPR
+            typename std::enable_if<
+                std::is_move_constructible<T>::value &&
+                std::is_move_assignable<T>::value
+             >::type
+            swap(T& x, T& y)
+            noexcept(std::is_nothrow_move_constructible<T>::value &&
+                     std::is_nothrow_move_assignable<T>::value)
+            {
+                T t(std::move(x));
+                x = std::move(y);
+                y = std::move(t);
+            }
```

Comment:
You are right. I've reverted the swap changes, and added this test to `test/utility/swap.cpp`.

After sleeping over it, I think the best thing is to just import std swap. Since ADL is used users can implement constexpr versions of swap for their own types if so desired. 

The only case that might be worth considering is that of C-arrays, but that can be done in the `swap_fn` object before performing ADL.